megamek/src/megamek/common/actions/WeaponAttackAction.java

```diff
@@ -662,3829 +629,4254 @@ private static ToHitData toHit(IGame game, int attackerId, Targetable target, in
-            toHit.addModifier(-1, Messages.getString("WeaponAttackAction.StableWeapon"));
+        // TAG can't be fired in any phase but offboard
+        if ((game.getPhase() != IGame.Phase.PHASE_OFFBOARD) && isTAG) {
+            return Messages.getString("WeaponAttackAction.TagOnlyInOffboard");
         }
         
-        if (weapon.hasQuirk(OptionsConstants.QUIRK_WEAP_NEG_MISREPAIRED)) {
-            toHit.addModifier(+1, Messages.getString("WeaponAttackAction.MisrepairedWeapon"));
-        }
+        // Unit-specific Reasons
         
-        if (weapon.hasQuirk(OptionsConstants.QUIRK_WEAP_NEG_MISREPLACED)) {
-            toHit.addModifier(+1, Messages.getString("WeaponAttackAction.MisreplacedWeapon"));
+        // Airborne units cannot tag and attack
+        // http://bg.battletech.com/forums/index.php?topic=17613.new;topicseen#new
+        if (ae.isAirborne() && ae.usedTag()) {
+            return Messages.getString("WeaponAttackAction.AeroCantTAGAndShoot");
         }
 
-        // Has the pilot the appropriate gunnery skill?
-        if (ae.hasAbility(OptionsConstants.UNOFF_GUNNERY_LASER)
-                && wtype.hasFlag(WeaponType.F_ENERGY)) {
-            toHit.addModifier(-1, Messages.getString("WeaponAttackAction.GunESkill"));
-        }
+        // Hull Down
 
-        if (ae.hasAbility(OptionsConstants.UNOFF_GUNNERY_BALLISTIC)
-                && wtype.hasFlag(WeaponType.F_BALLISTIC)) {
-            toHit.addModifier(-1, Messages.getString("WeaponAttackAction.GunBSkill"));
+        // Hull down mechs cannot fire any leg weapons
+        if (ae.isHullDown() && weapon != null) {
+            if (((ae instanceof BipedMech)
+                    && ((weapon.getLocation() == Mech.LOC_LLEG) || (weapon.getLocation() == Mech.LOC_RLEG)))
+                    || ((ae instanceof QuadMech) && ((weapon.getLocation() == Mech.LOC_LLEG)
+                            || (weapon.getLocation() == Mech.LOC_RLEG) || (weapon.getLocation() == Mech.LOC_LARM)
+                            || (weapon.getLocation() == Mech.LOC_RARM)))) {
+                return Messages.getString("WeaponAttackAction.NoLegHullDown");
+            }
         }
 
-        if (ae.hasAbility(OptionsConstants.UNOFF_GUNNERY_MISSILE)
-                && wtype.hasFlag(WeaponType.F_MISSILE)) {
-            toHit.addModifier(-1, Messages.getString("WeaponAttackAction.GunMSkill"));
+        // hull down vees can't fire front weapons
+        if ((ae instanceof Tank) && ae.isHullDown() && weapon != null && (weapon.getLocation() == Tank.LOC_FRONT)) {
+            return Messages.getString("WeaponAttackAction.FrontBlockedByTerrain");
         }
 
-        // Is the pilot a weapon specialist?
-        if (ae.hasAbility(OptionsConstants.GUNNERY_WEAPON_SPECIALIST, wtype.getName())) {
-            toHit.addModifier(-2, Messages.getString("WeaponAttackAction.WeaponSpec"));
-        } else if (ae.hasAbility(OptionsConstants.GUNNERY_SPECIALIST)) {
-			// aToW sty
```

Comment:
This might be a little out of scope, but I'm seeing a lot of these things where we loop through the game actions looking for a specific one. What if we were to introduce some kind of mechanism for setting a flag for these specific actions (perhaps a key-value pair with the key being an entity ID and the value being a set of flags indicating one of these special conditions such as "bomb action scheduled")? The only wrinkle would be that removing a particular action would have to unset that action's flag as well, but that should be doable as well.

Reply:
Yeah that's out of scope for this, but now things are organized so such improvements are more easily found and made.